aten/src/ATen/core/Tensor.h

```diff
@@ -1,3 +1,41 @@
 #pragma once
 
 #include <ATen/core/TensorBody.h>
+#include <c10/util/Exception.h>
+
+namespace at {
+class TORCH_API OptionalTensorRef {
+ public:
+  OptionalTensorRef() {}
+  OptionalTensorRef(const Tensor& src)
+      : ref_(c10::intrusive_ptr<TensorImpl>(
+            src.unsafeGetTensorImpl(),
+            c10::raw::DontIncreaseRefcount{})) {
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(src.defined());
+  }
+
+  OptionalTensorRef(const OptionalTensorRef& rhs)
```

Comment:
This isn't enough. You need to define copy assignment as well. And while we're at it, it's probably safest to explicitly define what the move constructor and move assignment do (though I think the defaults do what you want)

Reply:
(I think by manually defining the move constructor/assignment you can also apply a tiny optimization which is that you don't have to clear the source OptionalTensorRef on a move; it's non owning!)

Reply:
I added a move constructor/assignment operator but the generated assembly for a move assignment statement (see PR description) is exactly the same as what get I don't declare them at all. Interestingly enough, if I *do* declare them with `= default`, the assembly does change (and there are refcount manipulations involved).

It could be that I'm not feeding the right code into the compiler but I'm inclined to get rid of them since the implicitly generated ones seem to be doing the same thing as my implementations.

Reply:
Or maybe I implemented them incorrectly? I'm not clearing the source as far as I can tell...

Reply:
Hmm, so what it seems is happening is that when you explicitly declare copy constructor, this deletes the move constructor. So actually whenever a move would have been possible, the C++ compiler does a copy instead. Which normally would be a pessimization, but in this case it's fine! 

Prompted by this, I checked c++17 `string_view` to see what they'd done, and they had omitted them move assignment/constructor. So in fact you got it the right the first try and you should ignore me...

Reply:
(To be clear, you still need to define copy assignment; that's rule of three https://en.cppreference.com/w/cpp/language/rule_of_three )

Reply:
Okay, I deleted the move constructor and assignment operator.